workspaces/adr/plugins/adr/src/components/AdrReader/AdrReader.tsx

```diff
@@ -52,34 +54,81 @@ export const AdrReader = (props: {
     [adrFileLocationUrl],
   );
 
-  const adrContent = useMemo(() => {
+  const adrElements = useMemo(() => {
+    const elements: { reducedAdrContent: string; imageUrls: string[] } = {
+      reducedAdrContent: '',
+      imageUrls: [],
+    };
     if (!value?.data) {
-      return '';
+      return elements;
     }
     const adrDecorators = decorators ?? [
       adrDecoratorFactories.createRewriteRelativeLinksDecorator(),
       adrDecoratorFactories.createRewriteRelativeEmbedsDecorator(),
       adrDecoratorFactories.createFrontMatterFormatterDecorator(),
     ];
 
-    return adrDecorators.reduce(
+    elements.reducedAdrContent = adrDecorators.reduce(
       (content, decorator) =>
         decorator({ baseUrl: adrLocationUrl, content }).content,
       value.data,
     );
+    let imageUrlsArray = imageUrlsRegExp.exec(elements.reducedAdrContent);
+    while (imageUrlsArray !== null) {
+      elements.imageUrls.push(imageUrlsArray[2]);
+      imageUrlsArray = imageUrlsRegExp.exec(elements.reducedAdrContent);
+    }
+
+    return elements;
   }, [adrLocationUrl, decorators, value]);
 
+  const {
+    value: adrImagesBase64,
+    loading: adrImagesBase64Loading,
+    error: adrImagesBase64Error,
+  } = useAsync(async () => {
+    if (!adrElements?.imageUrls) {
+      return [];
+    }
+    const adrBase64Images = adrElements.imageUrls.map(imageUrl => {
+      return adrApi.imageAdr(imageUrl);
+    });
+    return await Promise.all(adrBase64Images);
+  }, [adrElements.imageUrls]);
+
+  const adrContent = useMemo(() => {
+    if (typeof adrImagesBase64 === 'undefined') {
+      return '';
+    }
+    let imageIterator = -1;
+    return adrElements.reducedAdrContent.replace(imageUrlsRegExp, (_, p1) => {
+      imageIterator += 1;
+      return `${p1}(${adrImagesBase64[imageIterator]?.data})`;
```

Comment:
Hm I think the previous solution you had using `transformImageUri` would be a cleaner solution here - I assume you ran into auth issues with the recent auth changes? That can be addressed by using the new `CookieAuthRefreshProvider` to leverage cookie auth here eg. TechDocs uses this to fetch embedded content with auth: https://github.com/backstage/backstage/blob/master/plugins/techdocs/src/reader/components/TechDocsReaderPage/TechDocsReaderPage.tsx#L192C7-L192C31

It should work here as well by wrapping the elements below with:
```tsx
<CookieAuthRefreshProvider pluginId="adr">
  <InfoCard>
    ...
  </InfoCard>
</CookieAuthRefreshProvider>

Reply:
Switched back to using transformImageUri.
Thank you for pointing out the `CookieAuthRefreshProvider`! it saved me :smile: